pkg/fieldpath/path.go

```diff
@@ -0,0 +1,243 @@
+//
+//  authShapeRef := &awssdkmodel.ShapeRef{
+//    ShapeName: "Author",
+//    Shape: &awssdkmodel.Shape{
+//      Type: "structure",
+//      MemberRefs: map[string]*awssdkmodel.ShapeRef{
+//        "Name": &awssdkmodel.ShapeRef{
+//          ShapeName: "Name",
+//          Shape: &awssdkmodel.Shape{
+//            Type: "string",
+//          },
+//        },
+//        "Address": &awssdkmodel.ShapeRef{
+//          ShapeName: "Address",
+//          Shape: &awssdkmodel.Shape{
+//            Type: "structure",
+//            MemberRefs: map[string]*awssdkmodel.ShapeRef{
+//              "State": &awssdkmodel.ShapeRef{
+//                ShapeName: "StateCode",
+//                Shape: &awssdkmodel.Shape{
+//                  Type: "string",
+//                },
+//              },
+//              "Country": &awssdkmodel.ShapeRef{
+//                ShapeName: "CountryCode",
+//                Shape: &awssdkmodel.Shape{
+//                  Type: "string",
+//                },
+//              },
+//            },
+//          },
+//        },
+//      },
+//    },
+//  }
+//
+// If I have the following Path:
+//
+// p := fieldpath.FromString("Author.Address.Country")
+//
+// calling p.ShapeRef(authShapeRef) would return the following:
+//
+// &awssdkmodel.ShapeRef{
+//   ShapeName: "CountryCode",
+//   Shape: &awssdkmodel.Shape{
+//     Type: "string",
+//   },
+// },
+func (p *Path) ShapeRef(
+	subject *awssdkmodel.ShapeRef,
+) *awssdkmodel.ShapeRef {
+	if subject == nil || p == nil || len(p.parts) == 0 {
+		return nil
+	}
+
+	// We first check that the first part in the path matches the supplied
+	// subject shape's name.
+	var compare *awssdkmodel.ShapeRef = subject
+	cp := p.Copy()
+	cur := cp.PopFront()
+	if compare.ShapeName != cur {
+		return nil
+	}
+	// And then we walk through the path, searching through the supplied
+	// ShapeRef for a member ShapeRef matching each path element.
+	for !cp.Empty() {
+		cur = cp.PopFront()
+		if compare = memberShapeRef(compare, cur); compare == nil {
+			return nil
+		}
+	}
+	return compare
+}
+
+// memberShapeRef returns the named member ShapeRef of the supplied
+// ShapeRef
+func memberShapeRef(
+	shapeRef *awssdkmodel.ShapeRef,
+	memberName string,
+) *awssdkmodel.ShapeRef {
+	if shapeRef.ShapeName == memberName {
+		return shapeRef
+	}
+	switch shapeRef.Shape.Type {
+	case "structure":
+		return shapeRef.Shape.MemberRefs[memberName]
+	case "list":
+		return memberShapeRef(&shapeRef.Shape.MemberRef, memberName)
+	case "map":
+		return memberShapeRef(&shapeRef.Shape.ValueRef, memberName)
+	}
+	return nil
+}
+
+// Contains returns true if the supplied string, delimited on ".", matches
+// p.parts up to the length of the supplied string.
+// e.g. if the Path p represents "A.B":
+//  subject "A" -> true
+//  subject "A.B" -> true
+//  subject "A.B.C" -> false
+//  subject "B" -> false
+//  subject "A.C" -> false
+func (p *Path) Contains(subject string) bool {
```

Comment:
nit: IMO "Contains" implies a substring search, however this is more like prefix search. How about naming the method "BeginsWith" ?

Reply:
> 

@vijtrip2 yeah, that's a good point. I grabbed this code from the runtime pkg/compare/path.go file and didn't really change it. I can change this to `[HasPrefix](https://pkg.go.dev/strings#HasPrefix)` to match the semantics of the strings package function of the same name?